client/src/main/java/io/pravega/client/stream/impl/EventStreamWriterImpl.java

```diff
@@ -302,25 +344,23 @@ public void flush() {
     }
     
     private void flushInternal() {
+        lock.readLock().lock();
         boolean success = false;
-        String sealedSegment = null;
-        while (!success) {
-            success = true;
-            synchronized (lock) {
+        try {
+            while (!success) {
+                success = true;
                 for (SegmentOutputStream writer : selector.getWriters()) {
                     try {
                         writer.flush();
                     } catch (SegmentSealedException e) {
-                        log.info("Segment was sealed during flush: {}", writer);
+                        //Segment sealed exception observed during a flush. Re-run flush on all the available writers.
```

Comment:
Writer.flush() is waiting for empty inflight events and when it  receives a segment sealed exception it implies that the segment sealed call back has executed and updated the writer list.
Hence a wait was not added.